data/transactions/verify/verifiedTxnCache.go

```diff
@@ -0,0 +1,295 @@
+		base:    0,
+	}
+	for i := 0; i < bucketsCount; i++ {
+		impl.buckets[i] = make(map[transactions.Txid]*GroupContext, entriesPerBucket)
+	}
+	return impl
+}
+
+// Add adds a given transaction group and it's associated group context to the cache. If any of the transactions already appear
+// in the cache, the new entry overrides the old one.
+func (v *verifiedTransactionCache) Add(txgroup []transactions.SignedTxn, groupCtx *GroupContext) {
+	v.bucketsLock.Lock()
+	defer v.bucketsLock.Unlock()
+	v.add(txgroup, groupCtx)
+}
+
+// AddPayset works in a similar way to Add, but is intended for adding an array of transaction groups, along with their corresponding contexts.
+func (v *verifiedTransactionCache) AddPayset(txgroup [][]transactions.SignedTxn, groupCtxs []*GroupContext) error {
+	v.bucketsLock.Lock()
+	defer v.bucketsLock.Unlock()
+	for i := range txgroup {
+		v.add(txgroup[i], groupCtxs[i])
+	}
+	return nil
+}
+
+// GetUnverifiedTranscationGroups compares the provided payset against the currently cached transactions and figure which transaction groups aren't fully cached.
+func (v *verifiedTransactionCache) GetUnverifiedTranscationGroups(txnGroups [][]transactions.SignedTxn, currSpecAddrs transactions.SpecialAddresses, currProto protocol.ConsensusVersion) (unverifiedGroups [][]transactions.SignedTxn) {
+	v.bucketsLock.Lock()
+	defer v.bucketsLock.Unlock()
+	groupCtx := &GroupContext{
+		specAddrs:        currSpecAddrs,
+		consensusVersion: currProto,
+	}
+	unverifiedGroups = make([][]transactions.SignedTxn, 0, len(txnGroups))
+
+	for txnGroupIndex := 0; txnGroupIndex < len(txnGroups); txnGroupIndex++ {
+		signedTxnGroup := txnGroups[txnGroupIndex]
+		verifiedTxn := 0
+		groupCtx.minTealVersion = logic.ComputeMinTealVersion(signedTxnGroup)
+
+		baseBucket := v.base
+		for txnIdx := 0; txnIdx < len(signedTxnGroup); txnIdx++ {
+			txn := &signedTxnGroup[txnIdx]
+			id := txn.Txn.ID()
+			// check pinned first
+			entryGroup := v.pinned[id]
+			// if not found in the pinned map, try to find in the verified buckets:
+			if entryGroup == nil {
+				// try to look in the previously verified buckets.
+				// we use the (base + W) % W trick here so we can go backward and wrap around the zero.
+				for offsetBucketIdx := baseBucket + len(v.buckets); offsetBucketIdx > baseBucket; offsetBucketIdx-- {
+					bucketIdx := offsetBucketIdx % len(v.buckets)
+					if params, has := v.buckets[bucketIdx][id]; has {
+						entryGroup = params
+						baseBucket = bucketIdx
+						break
+					}
+				}
+			}
+
+			if entryGroup == nil {
+				break
+			}
+
+			if !entryGroup.Equal(groupCtx) {
+				break
+			}
+
+			if entryGroup.signedGroupTxns[txnIdx].Sig != txn.Sig || (!entryGroup.signedGroupTxns[txnIdx].Msig.Equal(txn.Msig)) || (!entryGroup.signedGroupTxns[txnIdx].Lsig.Equal(&txn.Lsig)) || (entryGroup.signedGroupTxns[txnIdx].AuthAddr != txn.AuthAddr) {
+				break
+			}
+			verifiedTxn++
+		}
+		if verifiedTxn != len(signedTxnGroup) || verifiedTxn == 0 {
```

Comment:
why also append when verifiedTxn == 0?

Reply:
It's not a valid txn, and therefore would return in the list. I think it would never happen, but it's good to be on the safe side.